### DISCUSSION TITLE
Maintaining connections beyond the top level Async

### DISCUSSION THREAD

**Discussion by klobuczek** [ioquatix/ioquatix#14] [Q&A]
Hi Samuel,
I'm the maintainer of https://github.com/neo4jrb/neo4j-ruby-driver, a driver for the neo4j database. 
Neo4j Technologies provides official drivers for 5 languages and a testkit to ensure consistency of the driver API between languages. They have a synchronous and asynchronous api. 

neo4j-ruby-driver is an open-source driver and we have implemented so far the synchronous api. A simple example of the driver lifecycle can be found at https://github.com/neo4jrb/neo4j-ruby-driver/blob/90c3a861c5f23c1a0af2a7b44adabd7abfe99528/docs/dev_manual_examples.rb#L9-L23
We have had this driver implementation for a while using C or java extension. Currently, we are in the final stage of releasing a pure ruby driver. We are using async-io and async-pool.
 
Right now we hit a blocker. Our goal was to implement the synchronous api in a way so the client code does not have to define any async/sync demarkation. So we were creating a top-level Sync task with most of the client calls, but we noticed at the completion of every top-level task the pool with all connections is released. Is there a way of maintaining the pool with open connections across multiple reactors? Is there any way to implement the driver without the client controlling the async boundaries? Alternatively is there a way to implicitly set the top-level task around the entire client code/thread?
We would really appreciate your advice.

**Comment by ioquatix** [comment]
Thanks for your work and your questions.

The main reason why it is done this way is to prevent resources leaking. I'll try to give some context.

If your connection is a single IO, there is no reason why you can't have it outside of an event loop, but your users will need to close it correctly. Think of it as an implicit `with` block like in Python.

From a user point of view, every interaction should be discrete, i.e.

```
database.query(...) -> result
```

Internally, that checks out a connection, does the query, and gives the result.

If you need a long lived connection you can do something like this:

```
session = database.transaction
session.query(...) -> result
session.commit / abort
```

This gives you more explicit control over the life cycle.

Then, from the user's point of view, the connection pool is a cache of connections that can be used to amortise connection overhead and I should be clear, semantically that's the only reason why it exists. The connection pool in theory could just connect and disconnect for every resource acquire/release and it should not impact user code except for speed.

Regarding the life cycle of the connection pool, I think you have to expect that how connections are managed might not be completely trivial - as in - the policy for when to close connections and so on could be controlled by the user (e.g. pool minimum/maximum etc). Because pool resources are "async" resources, they can't exist outside of an event loop. Take for example, an HTTP/2 connection with multiplexing. It has a background reader task which is required to dispatch incoming frames to the right stream (multiplexed). Once you leave your asynchronous context (as formed by `Async`/`Sync`) that task must be terminated and thus the connection should be destroyed. The connection pool takes care of this. Regarding how that's used, it shouldn't impact code because you cannot assume the life cycle of connection pool resources and we see that behaviour as either user defined or implementation defined.

So, where this leaves you is: probably you should not close your connection in the connection pool. It means that the life cycle of your IO instance and probably your connection class is not determined by the life cycle of the pool.

```ruby
class DatabaseConnection
  def initialize(io, autoclose: true)
    @io = io
    @autoclose = autoclose
  end
  
  def close
    @io.close if @autoclose
  end
end
```

If you move a connection into the async event loop, you don't autoclose. In the end, it's up to you to determine what the right course of action is.

While not perfect, a good examle of what I consider the right model can be found in the db gem and the async-redis gem. Both follow the life-cycle of the async-pool.

In Ruby 3.x, normal IO becomes non-blocking when used in an asynchronous context, so in theory you don't need to do anything to get concurrent behaviour provided you use `IO` methods like `read`/`write`, `wait`, `wait_readable`,`wait_writable` etc which are all scheduling hooks. So you might find it easier to advise people to use that approach for connections that need to, say, be globally available. `async-pool` might be too limiting.

Just to make it crystal clear `async-pool` is designed to solve several key issues:
- Amortised connection cost (of course, main goal).
- Safe connection life cycle when going out of scope.
- Support unknown asynchronous scope and buffer connections as optimistically as possible.

> Right now we hit a blocker. Our goal was to implement the synchronous api in a way so the client code does not have to define any async/sync demarkation.

Fair enough. Ideally you don't need to care about this, but sometimes it's very hard because of multiplexing operations.

> So we were creating a top-level Sync task with most of the client calls, but we noticed at the completion of every top-level task the pool with all connections is released.

Yes this is expected behaviour.

> Is there a way of maintaining the pool with open connections across multiple reactors?

Not directly but you could use the autoclose approach above as one possible option. However your users might start leaking connections unexpectedly.

> Is there any way to implement the driver without the client controlling the async boundaries?

Yes, just use IO instances with your own wrapper and hope your users can deal with closing the connections at the right time.

> Alternatively is there a way to implicitly set the top-level task around the entire client code/thread?

This is actually how the fiber scheduler can work, but it's not the default approach since we retain compatibility with the same interface in Ruby 2.x. My goal is actually your driver code should not be concerned about async at all, just use normal blocking IO and other operations and you won't have to change anything.

I've given you quite a broad set of advice/ideas, so let me know if you want to discuss any specifics. In contrast, with the `db` gem, I'm very close to being able to run it both inside and outside of async with no changes, and it's my goal to support something like this but there are still a few rough edges. Ruby 3.2 would be the first version where I'm really happy to recommend it.
